
# SnpEff version 6

### Compatibility breaking

- Change interval to open
- Refactor:
  - [ ] Check all constructors
  - [ ] `setEnd()` -> `setEndClosed()`: Change code
  - [ ] `getEnd()` -> `getEndClosed()`: Change code
  - [ ] Repalce all direct access by getters / setters (getEndClosed(), setEndClosed())
  - [ ] New `setEnd()`

### Splice variants
- [ ] TestCasesIntegrationSpliceRegion: SNV (e.g. compare to ensembl)
- [ ] TestCasesIntegrationSpliceRegion: INS (e.g. compare to ensembl)
- [ ] TestCasesIntegrationSpliceRegion: DEL. Add test case, (compare?)

# AlphaFold

- [ ] Work on test class Zzz.java
- [ ] Find a way to get hydrogen bonds and other (covalent, etc)

# Refactoring

- [ ] Make Interval.start / end private
- [ ] Replace access by getEndClosed  / setEndClosed
- [ ] Try to remove getEndClosed  / setEndClosed, as much as possible
- [ ] Make sure serializations use getEnd (not getEndClosed)[
- [ ] Make sure toString() uses getEnd() instead of getEndClosed() ???

### Field: end

#### Value read
SnpEff
org.snpeff.interval
-[ ] Cds
     toString()
     		return getChromosomeName() + "\t" + start + "-" + end //
-[ ] Chromosome
     toString()
     		return getChromosomeName() + "\t" + start + "-" + end //
-[ ] Custom
     toString()
     		sb.append(end);
-[ ] Downstream
     distanceToTr(Variant)
     		int dist = (parent.isStrandPlus() ? variant.getStart() - start : end - variant.getStart()) + 1;
-[ ] Exon
     createSpliceSiteRegionEnd(int)
     		if (isStrandPlus()) spliceSiteRegionEnd = new SpliceSiteRegion(this, end - (size - 1), end, strandMinus, id);
     createSpliceSiteRegionStart(int)
     		else spliceSiteRegionStart = new SpliceSiteRegion(this, end - (size - 1), end, strandMinus, id);
     sanityCheck(Variant)
     		int mend = Math.min(variant.getEnd(), end);
     toString()
     			return getChromosomeName() + ":" + start + "-" + end //
     			return getChromosomeName() + ":" + start + "-" + end //
-[ ] Gene
     adjust()
     			if (end != newEnd) {
     				if (Config.get().isVerbose()) Log.warning(ErrorWarningType.WARNING_GENE_COORDINATES, "Gene '" + id + "' (name:'" + geneName + "'), adjusting end coordinate from " + end + " to " + newEnd);
     circularClone()
     		} else if (end > chr.getEnd()) {
     toString(boolean)
     		sb.append(getChromosomeName() + ":" + start + "-" + end);
-[ ] GffMarker
     parseId()
     		else id = gffType + "_" + getChromosomeName() + "_" + (start + 1) + "_" + (end + 1); // No ID => create one
     toString()
     				+ "\t" + end //
-[ ] Interval
     compareTo(Interval)
             if (end > i2.end) return 1;
             if (end < i2.end) return -1;
     getEnd()
             return end;
     hashCode()
             hashCode = hashCode * 31 + end;
     intersects(int, int)
             return (iEnd >= start) && (iStart <= end);
     intersects(Interval)
             return (interval.getEnd() >= start) && (interval.getStart() <= end);
     intersects(long)
             return (start <= point) && (point <= end);
     intersects(Marker)
             return (interval.getEnd() >= start) && (interval.getStart() <= end);
     intersectSize(Marker)
             int end = Math.min(this.end, interval.getEnd());
     isCircular()
                     || (start > end) // Start before end?
                     || (end > chr.getEnd()) // Ends after chromosome end?
     isValid()
             return start <= end;
     size()
             return end - start + 1;
     toStr()
                     + "-" + (end + 1) //
     toString()
             return start + "-" + end //
     toStringAsciiArt(int)
                 if ((i >= start) && (i <= end)) sb.append('-');
     toStrPos()
             return getChromosomeName() + ":" + (start + 1) + "-" + (end + 1);
     Intron
     createSpliceSiteAcceptor(int)
     			ssstart = end - (maxSpliceSiteSize - 1);
     			ssend = end;
     createSpliceSiteDonor(int)
     			ssstart = end - (maxSpliceSiteSize - 1);
     			ssend = end;
     createSpliceSiteRegionEnd(int, int)
     		if (isStrandPlus()) spliceSiteRegionEnd = new SpliceSiteRegion(this, end - (sizeMax - 1), end - (sizeMin - 1), strandMinus, id);
     createSpliceSiteRegionStart(int, int)
     		else spliceSiteRegionStart = new SpliceSiteRegion(this, end - (sizeMax - 1), end - (sizeMin - 1), strandMinus, id);
-[ ] Marker
     adjust(Marker)
             end = Math.max(end, child.getEnd());
     applyDel(Variant)
                 if (m.end < variant.getEnd()) {
                 int iend = Math.min(variant.getEnd(), m.end);
     applyIns(Variant)
             } else if (variant.getStart() <= m.end) {
     cloneShallow()
                 clone.end = end;
     compareToPos(Interval)
             if (end > i2.end) return 1;
             if (end < i2.end) return -1;
     distance(Marker)
             if (interval.getStart() > end) return interval.getStart() - end;
     includes(Marker)
             return (start <= marker.start) && (marker.end <= end);
     intersect(Marker)
             int iend = Math.min(end, marker.getEnd());
     serializeSave(MarkerSerializer)
                     + "\t" + end //
     shouldApply(Variant)
             return variant.getStart() <= end;
     toString()
             return getChromosomeName() + "\t" + start + "-" + end //
     union(Marker)
             int uend = Math.max(end, m.getEnd());
-[ ] Markers
     merge()
                 end = Math.max(end, i.end);
     toStringAsciiArt(int)
                 sb.append("|" + i.toStringAsciiArt(maxLen) + "|\t" + i.getChromosomeName() + ": [" + i.start + " - " + i.end + "] ");
-[ ] MarkerSeq
     toString()
     		return getChromosomeName() + ":" + start + "-" + end //
     union(Marker)
     		int uend = Math.max(end, m.getEnd());
     				int overlap = end - m.start + 1;
     				int overlap = m.end - start + 1;
-[ ] Motif
     effectImpact(Variant)
     			MarkerSeq mseq = new MarkerSeq((Marker) parent, start, end, false, id); // Notice: We use positive strand
-[ ] Regulation
     toString()
     		return getChromosomeName() + "\t" + start + "-" + end //
-[ ] SpliceSiteAcceptor
     intersectsCoreSpliceSite(Marker)
     			coreEnd = end;
-[ ] SpliceSiteDonor
     intersectsCoreSpliceSite(Marker)
     			coreEnd = end;
-[ ] Transcript
     adjust()
                 if (end != newEnd) {
     calcCdsStartEnd()
                     cdsStart = (isStrandPlus() ? end : start); // cdsStart is the position of the first base in the CDS (i.e. the first base after all 5'UTR)
                     cdsEnd = (isStrandPlus() ? start : end); // cdsEnd is the position of the last base in the CDS (i.e. the first base before all 3'UTR)
                     cdsStart = (isStrandPlus() ? start : end); // cdsStart is the position of the first base in the CDS (i.e. the first base after all 5'UTR)
                     cdsEnd = (isStrandPlus() ? end : start); // cdsEnd is the position of the last base in the CDS (i.e. the first base before all 3'UTR)
     createUpDownStream(int)
             int afterStart = Math.min(end + 1, chrMax);
             int afterEnd = Math.min(end + upDownLength, chrMax);
     toString(boolean)
             sb.append(getChromosomeName() + ":" + start + "-" + end);
     toStringAsciiArt(boolean)
             for (int i = start, j = 0; i <= end; i++, j++) {
             for (int i = start; i <= end; i++) {
-[ ] Upstream
     distanceToTr(Variant)
     		int dist = (parent.isStrandPlus() ? end - variant.getStart() : variant.getStart() - start) + 1;
-[ ] Variant
     hashCode()
             hashCode = hashCode * 31 + end;
     lengthChange()
             return end - start;
     netChange(Marker)
             int removeAfter = end - marker.getEnd();
     toString()
             ) return "chr" + getChromosomeName() + ":" + start + "-" + end + "[" + variantType + "]";
     toStringEnsembl()
             return getChromosomeName() + "\t" + start + "\t" + end + "\t" + ref + "/" + alt + "\t+";
     org.snpeff.vcf
-[ ] VcfEntry
     parseEnd(String)
     				if (end < start) { throw new RuntimeException("INFO field 'END' is before varaint's 'POS'\n\tEND : " + end + "\n\tPOS : " + start); }
     variants(Chromosome, int, String, String, String)
     				if (end > start) {
     					int size = end - startNew + 1;
     				Variant var = new Variant(chromo, startNew, end, id);
     				Variant var = new Variant(chromo, startNew, end, id);
#### Value write
     SnpEff
     org.snpeff.interval
-[ ] Cds
     frameCorrection(int)
     		else end -= frameCorrection;
-[ ] Chromosome
     setLength(int)
     		end = len - 1; // Intervals are zero-based
-[ ] Exon
     frameCorrection(int)
     		else end -= frameCorrection;
-[ ] Gene
     adjust()
     				end = newEnd;
-[ ] GffMarker
     parse(Genome, String)
     		end = Gpr.parseIntSafe(fields[4]) - 1;
-[ ] Interval
     Interval()
             end = -1;
     Interval(Interval, int, int, boolean, String)
             this.end = end;
     setEnd(int)
             this.end = end;
     shiftCoordinates(int)
             end += shift;
-[ ] Marker
     adjust(Marker)
             end = Math.max(end, child.getEnd());
     applyDel(Variant)
                 m.end += lenChange;
                 m.end += variant.lengthChange();
                 m.end -= (iend - istart + 1); // Update end coordinate
                     m.end -= delta;
     applyDup(Variant)
                 m.end += lenChange;
                 m.end += lenChange;
                 m.end += variant.lengthChange();
                 m.end += variant.intersect(m).size();
     applyIns(Variant)
                 m.end += lenChange;
                 m.end += variant.lengthChange();
     cloneShallow()
                 clone.end = end;
     readTxt(String, int, Genome, int)
                     end = Gpr.parseIntSafe(fields[2]) - positionBase;
     serializeParse(MarkerSerializer)
             end = markerSerializer.getNextFieldInt();
-[ ] MarkerSeq
     union(Marker)
     		msNew.end = uend;
-[ ] Variant
     init(Marker, int, String, String, VariantType, String)
                 end = start;
                 if (ref.length() > 1) end = start + ref.length() - 1;
     org.snpeff.vcf
-[ ] VcfEntry
     parseEnd(String)
     		end = start + ref.length() - 1;
     				end = (int) getInfoInt(VCF_INFO_END) - 1;
